itest/lndsigner_test.go

```diff
@@ -0,0 +1,840 @@
+// addresses
+func (tctx *testContext) testSweepToP2WKH(t *testing.T) {
+	tctx.lndMtx.RLock()
+	defer tctx.lndMtx.RUnlock()
+
+	for _, lnd := range tctx.lnds {
+		resp := lnd.Lncli("newaddress", "p2wkh")
+		address := resp["address"].(string)
+
+		resp = lnd.Lncli("sendcoins", "--sweepall",
+			address)
+		require.Equal(t, 64, len(resp["txid"].(string)))
+
+		t.Logf("%s", lnd.idPubKey)
+	}
+}
+
+// testSweepToNP2WKH sweeps all of the nodes' on-chain funds into NP2WKH
+// addresses
+func (tctx *testContext) testSweepToNP2WKH(t *testing.T) {
+	tctx.lndMtx.RLock()
+	defer tctx.lndMtx.RUnlock()
+
+	for _, lnd := range tctx.lnds {
+		resp := lnd.Lncli("newaddress", "np2wkh")
+		address := resp["address"].(string)
+
+		resp = lnd.Lncli("sendcoins", "--sweepall",
+			address)
+		require.Equal(t, 64, len(resp["txid"].(string)))
+
+		t.Logf("%s", lnd.idPubKey)
+	}
+}
+
+// testSweepToP2TR sweeps all of the nodes' on-chain funds into P2TR
+// addresses
+func (tctx *testContext) testSweepToP2TR(t *testing.T) {
+	tctx.lndMtx.RLock()
+	defer tctx.lndMtx.RUnlock()
+
+	for _, lnd := range tctx.lnds {
+		resp := lnd.Lncli("newaddress", "p2tr")
+		address := resp["address"].(string)
+
+		resp = lnd.Lncli("sendcoins", "--sweepall",
+			address)
+		require.Equal(t, 64, len(resp["txid"].(string)))
+
+		t.Logf("%s", lnd.idPubKey)
+	}
+}
+
+// testEachPaysEachOther sends LN payments from each LND to each other LND,
+// testing both direct and chained payments.
+func (tctx *testContext) testEachPaysEachOther(t *testing.T) {
+	tctx.testEachPair(func(lnd1, lnd2 *lndHarness) {
+		resp := lnd1.Lncli("addinvoice", "5000")
+		invoice := resp["payment_request"].(string)
+
+		resp = lnd2.Lncli("payinvoice", "--timeout=10s", "--json",
+			"-f", invoice)
+		require.Equal(t, resp["status"].(string), "SUCCEEDED")
+
+		t.Logf("%s paid %s", lnd2.idPubKey,
+			lnd1.idPubKey)
+	})
+}
+
+// testEachSignVerifyEachOther signs a message from each LND to each other LND,
+// verifying the message on the second LND.
+func (tctx *testContext) testEachSignVerifyEachOther(t *testing.T) {
+	tctx.testEachPair(func(lnd1, lnd2 *lndHarness) {
+		message := lnd1.idPubKey + " to " + lnd2.idPubKey
+
+		resp := lnd1.Lncli("signmessage", message)
+		sig := resp["signature"].(string)
+
+		resp = lnd2.Lncli("verifymessage", message, sig)
+		require.True(t, resp["valid"].(bool))
+
+		t.Logf(message)
+	})
+}
+
+// testContext manages the test environment.
+type testContext struct {
+	t      *testing.T
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	tmpRoot string
+
+	vaultPort   string
+	vaultCmd    *exec.Cmd
+	vaultMtx    sync.RWMutex
+	vaultClient *api.Logical
+
+	bitcoinDir     string
+	bitcoinRPC     string
+	bitcoinZB      *net.TCPAddr
+	bitcoinZT      *net.TCPAddr
+	bitcoindCmd    *exec.Cmd
+	bitcoindMtx    sync.RWMutex
+	bitcoindClient *api.Logical
+
+	lndPath        string
+	lndSignerPath  string
+	lncliPath      string
+	bitcoincliPath string
+
+	lndMtx sync.RWMutex
```

Comment:
The tests still give me a race condition because these accesses happen from different goroutines without  being gated by a mutex, even if it's technically not possible that it happens at the same time.

Reply:
I believe I've fixed this in https://github.com/bottlepay/lndsigner/compare/d7ba0804a5348e157f205542fa403bf371d91d80..470890621d664c50ed3baafabefc2b0921027a69

Reply:
Looks like passing `t` into `waitProc` is a bad idea. Will fix.

Reply:
So the error was caused by calling `t.Logf` with a saved `t` all along. Fixed in https://github.com/bottlepay/lndsigner/compare/b1998d79e511c3bf6ceecdf829d711b6323fe512..b03297c3e6a00d67a20495218d804f98dc721d86 by replacing `zap`'s test logger with a development logger that doesn't rely on a `*testing.T`.

However, this is a good sign to stop `lnd` and `bitcoind` via RPC instead of OS signal.

Reply:
I've replaced the OS interrupt signal to `lnd` and `bitcoind` with a `lncli stop` and `bitcoin-cli stop`, which still ends up with the processes being killed as the main test exits, and blurting out a warning. I silenced it on `signal: killed` errors now, so if the daemons are too slow to shut down after the tests, we don't need to hear about it.

Reply:
>So the error was caused by calling t.Logf with a saved t all along.

What was the root cause of this failing? `t` being used in a goroutine? I think that should be possible.